### ISSUE TITLE
DATEFORMAT issue in RedshiftWriter.scala

### ISSUE THREAD

**Issue by jackar** [databricks/spark-redshift#122]
There's no need for a dateformat - especially since dates don't have millisecond level granularity, that would be timeformat, but that's already the default...

http://docs.aws.amazon.com/redshift/latest/dg/r_DATEFORMAT_and_TIMEFORMAT_strings.html

````
private def copySql(
      sqlContext: SQLContext,
      params: MergedParameters,
      creds: AWSCredentials,
      manifestUrl: String): String = {
    val credsString: String = AWSCredentialsUtils.getRedshiftCredentialsString(creds)
    val fixedUrl = Utils.fixS3Url(manifestUrl)
    s"COPY ${params.table.get} FROM '$fixedUrl' CREDENTIALS '$credsString' FORMAT AS " +
      s"AVRO 'auto' DATEFORMAT 'YYYY-MM-DD HH:MI:SS' manifest ${params.extraCopyOptions}"
  }
````

"DATEFORMAT 'YYYY-MM-DD HH:MI:SS' " should be removed from Line 101

I found this issue when reading a parquet file from s3 and then trying to save the dataframe into redshift:

````
Error (code 1205) while loading data into Redshift: "Date value did not match format specified [YYYY-MM-DD HH:MI:SS]"
````

The string format for date I'm using is YYYY-MM-DD

**Comment by JoshRosen** [comment]
Hi @jackar,

To provide a bit of historical context: `spark-avro` does not support date or timestamp columns, so in order to generate the unload data we end up converting dates and timestamps into strings prior to loading them into Redshift. See #39 for a discussion of why we use this particular string format, vs. other formats.

To help me understand the error message that you saw, can you share the schema of the DataFrame that you were saving back to Redshift? If you have a string column in YYYY-MM-DD format and are relying on Redshift to perform the conversion, then I don't think that's going to be compatible with the formats that we need to set in order to handle timestamp conversion. In that case, my recommendation would be to use Spark's built-in date and time functions to convert the YYYY-MM-DD  into a SQL date prior to saving to Redshift.

**Comment by JoshRosen** [comment]
According to the [Redshift documentation](https://docs.aws.amazon.com/redshift/latest/dg/r_Examples_with_datetime_types.html):

> If you attempt to insert a timestamp value into a DATE column, the time literal is truncated and the date itself is loaded.

After a closer look at the Redshift docs and #39, I'm now a bit confused about why we unloaded the DATE column as a timestamp in the first place. I'd be fine with switching to the default `DATEFORMAT` (`YYYY-MM-DD`) and updating the tests. @jackar, you're welcome to submit a PR for this and I'll be able to help review. You'll need to update the code which writes the Avro files so that dates are encoded into the correct format; you may also need to update some unit tests.

**Comment by jackar** [comment]
Thanks for the quick response Josh. I'll investigate soon, but yep, I do have a string column in YYYY-MM-DD format coming from a parquet file that I wrote to s3 using Spark, but it looks like after I cast the timestamp to a date 

````
df.withColumn("ymd", df("timestamp").cast(DateType)
````

and get

````
|-- ymd: date (nullable = true)
````

in the schema and then save said dataframe to a parquet file is when the implicit conversion to string occurs because when I read the s3 parquet file and check the schema again, I get the following

````
|-- ymd: string (nullable = true)
````

I imagine that spark-avro is a dependency w.r.t. spark-parquet as you've described, which makes sense at to why the date type is missing. But it seems like you're referring to spark-redshift itself when you say that the implicit conversions happen, not parquet. After I make the fix to DATEFORMAT I'll try see if I need to cast the ymd string back into a date.

I don't think I'll have time today, but I'd be really happy to open the PR later this week.

**Comment by JoshRosen** [comment]
Yeah, the implicit conversions that I mentioned take place inside of `spark-redshift`'s own write path.

I'm fairly sure that Spark's built-in Parquet support does support date columns. Spark's built-in Parquet support does not depend on the `spark-avro` library.

I tried it out in the shell:

```scala
scala> val df = sc.parallelize(Seq(("Name", "2015-10-18"))).toDF("name", "date")
df: org.apache.spark.sql.DataFrame = [name: string, date: string]

scala> import org.apache.spark.sql.types._
import org.apache.spark.sql.types._

scala> val df2 = df.withColumn("date", df("date").cast(DateType))
df2: org.apache.spark.sql.DataFrame = [name: string, date: date]

scala> df2.write.parquet("df2")
[...]

scala> sqlContext.read.parquet("df2").printSchema
root
 |-- name: string (nullable = true)
 |-- date: date (nullable = true)
```

Are you sure that you actually saved the transformed Dataframe with the casted column instead of saving the original DF?

**Comment by jackar** [comment]
I finally had some time to make the fix and open a PR, but let me know what you think, hopefully I didn't miss anything nuanced. 

And regarding DateType... I think I found the underlying issue. After writing to s3 and partitioning by a DateType column, reads on the parquet "table" (i.e. s3n://s3_bucket_url/table where date partitions break the table into date-based s3n://s3_bucket_url/table/date=2015-11-25 chunks)  will show the partitioned date column as a StringType. 

To test this hypothesis I created two date columns where written data is partitioned by one, as follows:

````
val df1 = df0.withColumn("date0", df0("date").cast(DateType))
val df2 = df1.withColumn("date1", df1("date").cast(DateType))
// keeping hour in because I want to include accurate info regarding my test case
df2.write.mode("append").partitionBy("date", "hour").parquet(s"s3n://<s3_bucket_url>/table")
````

A simple workaround is to create two date columns or partition by year, month, day (i.e. /year=2015/month=11/day=25). I imagine date partitioning is a common use case so it might make sense to support the feature. Thoughts?

Link to similar issue for presto and hive tables partitioned by date:
https://github.com/facebook/presto/issues/3786

**Comment by JoshRosen** [comment]
The Parquet issue seems like a Spark problem, not a `spark-redshift`-specific issue, so do you mind filing an ticket at https://issues.apache.org/jira/browse/SPARK to investigate / track that? I'll ping some folks who are more familiar with Hive partitioned tables to help investigate.

**Comment by jackar** [comment]
Sure, I'll include a link for future reference: 

https://issues.apache.org/jira/browse/SPARK-11995

**Comment by JoshRosen** [comment]
Great! I'm going to close this issue, since it was fixed by your patch (#130).
